feat(onboarding): Show JS Loader in onboarding

mydea opened the pull request:
Follow up/fork of https://github.com/getsentry/sentry/pull/46452

This shows a new UI in the new onboarding for the JS Loader. This is shown when the platform is plain browser JS.

![Screenshot 2023-03-30 at 12 21 31](https://user-images.githubusercontent.com/2411343/228818637-617ca55a-aba7-48e6-99ef-0f8a9384757f.png)

![Screenshot 2023-03-30 at 12 21 36](https://user-images.githubusercontent.com/2411343/228818648-4d6568b8-ecf5-4191-8f16-eeb38f301956.png)

Notes on the implementation:

* We update the loader settings automatically whenever the user toggles a checkbox
* We _do not_ consider the settings that have been present on the DSN - as we always want to overwrite this with the defaults.
* We assume that a project has a single DSN setup and use this

Closes https://github.com/getsentry/sentry/issues/46718

matejminar reviewed (approved):
Looks good to me.

matejminar commented:
@mydea I think we are missing the "Prefer to set up Sentry using NPM or Yarn? Go here." alert from Figma. Is that out of the scope of the initial test?

mydea commented:
So I updated this and added a skip button: 

![image](https://user-images.githubusercontent.com/2411343/229083243-cf1283fd-5614-421e-94e7-dd55458fbd4f.png)

mydea committed:
feat(onboarding): Add project loader UI

mydea committed:
add trest

mydea commented:
OK, so I added some basic tests for this as well, should be good to go now from my POV!

evanpurkhiser commented:
Feel free to merge, but I would like to leave some comments on this in the morning (PST)

I’m interested in what our long term plan is for dealing with highly specific documentation components like this. Typically this imports documentation from the docs website. But this component does not seem to do that (correct me if I am wrong)

priscilawebdev commented on static/app/components/onboarding/productSelection.tsx:
```diff
@@ -19,9 +21,15 @@ export enum PRODUCT {
 
 type Props = {
   defaultSelectedProducts?: PRODUCT[];
+  lazyLoader?: boolean;
+  skipLazyLoader?: () => void;
```
what if we only use the `skipLazyLoader` for the check? if it is provided we render everything we are currently rendering for `lazyLoader` otherwise the default version

mydea replied:
hmm, not sure, feels a bit less explicit to me, but happy to do however you feel works best!

priscilawebdev replied:
we can do this in  a follow-up 😉

priscilawebdev commented on static/app/views/onboarding/setupDocs.tsx:
```diff
@@ -432,6 +447,14 @@ function SetupDocs({search, route, router, location, ...props}: Props) {
+              project={project}
+              location={location}
+              platform={loadedPlatform}
+              close={() => setShowLoaderOnboarding(false)}
```
I like this solution, but we might consider saving it to localStorage in the future for the case users decide to reload the page

priscilawebdev commented on static/app/views/onboarding/setupDocsLoader.tsx:
```diff
@@ -0,0 +1,240 @@
+      {projectKeyUpdateError && (
+        <LoadingError
+          message={t('Failed to update the project key with the selected products.')}
+          onRetry={() => updateSelectedProducts()}
```
```suggestion
          onRetry={updateSelectedProducts}
```

nit: shall we name it `handleUpdateSelectedProducts`?

mydea replied:
makes sense! 👍

priscilawebdev commented on static/app/views/onboarding/setupDocsLoader.tsx:
```diff
@@ -0,0 +1,240 @@
+  const updateSelectedProducts = useCallback(async () => {
+    const productsQuery =
+      (location.query.product as PRODUCT | PRODUCT[] | undefined) ?? [];
+    const products = typeof productsQuery === 'string' ? [productsQuery] : productsQuery;
```
we can use the function https://github.com/getsentry/sentry/blob/7dd5aa6ba15a703fad25766356dccf4018ed6a03/static/app/utils/queryString.tsx#L98 here 😉

priscilawebdev commented on static/app/views/onboarding/setupDocsLoader.tsx:
```diff
@@ -0,0 +1,240 @@
+      return;
+    }
+
+    const apiEndpoint = `/projects/${organization.slug}/${project.slug}/keys/${keyId}/`;
```
```suggestion
```

priscilawebdev commented on static/app/views/onboarding/setupDocsLoader.tsx:
```diff
@@ -0,0 +1,240 @@
+    });
+
+    try {
+      await api.requestPromise(apiEndpoint, {
```
```suggestion
      await api.requestPromise(`/projects/${organization.slug}/${project.slug}/keys/${keyId}/`, {
```

priscilawebdev commented on static/app/views/onboarding/setupDocsLoader.tsx:
```diff
@@ -0,0 +1,240 @@
+
+    const apiEndpoint = `/projects/${organization.slug}/${project.slug}/keys/${keyId}/`;
+
+    const newDynamicSdkLoaderOptions: Record<DynamicSDKLoaderOption, boolean> = {
```
this could be outside of the component

priscilawebdev commented on static/app/views/onboarding/setupDocsLoader.tsx:
```diff
@@ -0,0 +1,240 @@
+        </OptionalConfigWrapper>
+        {showOptionalConfig && (
+          <div>
+            <CodeSnippet dark language="js" onCopy={() => {}}>
```
shall we update the `onCopy` function? currently it won't do anything

mydea replied:
I removed this!

priscilawebdev replied:
I find it quite useful... maybe we can follow up on this in another PR? 🙏

priscilawebdev commented:
@evanpurkhiser we are going to work on a [Spike](https://github.com/orgs/getsentry/projects/79?pane=issue&itemId=24581836) for a more scalable solution for the docs and one of our ideas is to have the docs we load on the onboarding in our sentry repo, but for now we are just doing experiments and in case sticking to it, a more formal proposal will be brought to Frontend TSC

priscilawebdev commented on static/app/views/onboarding/setupDocs.tsx:
```diff
@@ -263,10 +264,14 @@ function SetupDocs({search, route, router, location, ...props}: Props) {
 
+  const loaderOnboarding = !!organization.features?.includes('onboarding-project-loader');
+
+  const jsDynamicLoader = !!organization.features?.includes('js-sdk-dynamic-loader');
```
I think actually we could have used only one feature flag 🤔

mydea replied:
should I remove the other flag and do it based on this? But maybe it is still OK to leave this as I _think_ we'll flip the `js-sdk-dynamic-loader` one on by default soon!

mydea committed:
PR feedback

mydea committed:
PR fixes

priscilawebdev commented:
![image](https://user-images.githubusercontent.com/29228205/229528071-59568de4-8ffc-4e89-ae8c-0f052e3509ef.png)
we don't have the dismiss `x` anymore in the blue alert ... do we still plan to implement that?

priscilawebdev commented:
something for the future...we should think of a away for users to go back to the 'loader script' because currently, he can't go back to the previous UI 

https://user-images.githubusercontent.com/29228205/229528946-a85a0f87-65e9-4d7f-90a5-97ff1fb4fb55.mov

priscilawebdev reviewed (approved):
left some comments that are not blockers... well done 🎉 🥔

mydea committed:
Update static/app/views/onboarding/setupDocsLoader.tsx

Co-authored-by: Priscila Oliveira <priscila.oliveira@sentry.io>

evanpurkhiser commented on static/app/components/onboarding/productSelection.tsx:
```diff
@@ -135,6 +147,19 @@ export function ProductSelection({defaultSelectedProducts}: Props) {
       </Products>
+      {lazyLoader && (
+        <AlternativeInstallationAlert type="info" showIcon>
+          {tct('Prefer to set up Sentry using [npm] or [yarn]? [goHere].', {
+            npm: <strong>npm</strong>,
+            yarn: <strong>yarn</strong>,
+            goHere: (
+              <Button onClick={skipLazyLoader} priority="link">
+                {t('Go here')}
+              </Button>
+            ),
```
```suggestion
          {tct('Prefer to set up Sentry using [npm:npm] or [yarn:yarn]? [goHere:Go here].', {
            npm: <strong />,
            yarn: <strong />,
            goHere: (
              <Button onClick={skipLazyLoader} priority="link" />
            ),
```

priscilawebdev replied:
hm, I like the suggestion with the exception for the button... will it have the proper `aria-label` and `children`?

mydea committed:
Update static/app/components/onboarding/productSelection.tsx

Co-authored-by: Evan Purkhiser <evanpurkhiser@gmail.com>

getsantry[bot] committed:
style(lint): Auto commit lint changes

mydea committed:
update instructions

mydea commented:
Updated the instructions at the bottom:

![image](https://user-images.githubusercontent.com/2411343/229747803-181fd5c1-e612-40fe-9402-8e8351c9acb1.png)

Will make a follow up PR that actually sets the `firstEvent` when the loader is loaded!

mydea committed:
UI fixes

mydea committed:
style improvements

mydea commented:
Slight UI improvements based on feedback by @Jesse-Box 

![image](https://user-images.githubusercontent.com/2411343/229794633-e8a6212e-e047-4504-8ff9-f957bd47bae1.png)

mydea committed:
add script tag
